apps/errors/error_design.md

```diff
@@ -0,0 +1,119 @@
+
+## Background
+
+Errors in MSAL are intended for app developers to troubleshoot and not for displaying to end-users.
+
+### Go error handling vs other MSAL languages
+
+Most modern languages use exception based errors. Simply put, you "throw" an exception and it must be caught at some routine in the upper stack or it will eventually crash the program.
+
+Go doesn't use exceptions, instead it relies on multiple return values, one of which can be the builtin error interface type. It is up to the user to decide what to do.
+
+### Go custom error types
+
+Errors can be created in Go by simply using errors.New() or fmt.Errorf() to create an "error".
+
+Custom errors can be created in multiple ways. One of the more robust ways is simply to satisfy the error interface:
+
+```go
+type MyCustomErr struct {
+  Msg string
+}
+func (m MyCustomErr) Error() string { // This implements "error"
+  return m.Msg
+}
+```
+
+### MSAL Error Goals
+
+- Provide diagnostics to the user and for tickets that can be used to track down bugs or client misconfigurations
+- Detect errors that are transitory and can be retried
+- Allow the user to identify certain errors that the program can respond to, such a informing the user for the need to do an enrollment
+  
+## Implementing Client Side Errors
+
+Client side errors indicate a misconfiguration or passing of bad arguments that is non-recoverable. Retrying isn't possible.
+
+These errors can simply be standard Go errors created by errors.New() or fmt.Errorf(). If down the line we need a custom error, we can introduce it, but for now the error messages just need to be clear on what the issue was.
+
+## Implementing Service Side Errors
+
+Service side errors occur when an external RPC responds either with an HTTP error code or returns a message that includes an error.
+
+These errors can be transitory (please slow down) or permanent (HTTP 404).  To provide our diagnostic goals, we require the ability to differentiate these errors from other errors.
+
+The current implementation includes a specialized type that captures any error from the server:
+
+```go
+// CallErr represents an HTTP call error. Has a Verbose() method that allows getting the
+// http.Request and Response objects. Implements error.
+type CallErr struct {
+    Req  *http.Request
+    Resp *http.Response
+    Err  error
+}
+
+// Errors implements error.Error().
+func (e CallErr) Error() string {
+    return e.Err.Error()
+}
+
+// Verbose prints a versbose error message with the request or response.
+func (e CallErr) Verbose() string {
+    e.Resp.Request = nil // This brings in a bunch of TLS crap we don't need
+    e.Resp.TLS = nil     // Same
+    return fmt.Sprintf("%s:\nRequest:\n%s\nResponse:\n%s", e.Err, prettyConf.Sprint(e.Req), prettyConf.Sprint(e.Resp))
+}
+```
+
+A user will always receive the most concise error we provide.  They can tell if it is a server side error using the Go error package:  
+
+```go
+if errors.Is(err, CallErr{}) {
```

Comment:
This is incorrect.  `errors.Is()` is for checking for a sentinel error value like `io.EOF`.

https://blog.golang.org/go1.13-errors

Reply:
Not a Go expert, but I think you are right. 
From what the blog says,  we can still use errors.As() for type checking to see if the errors is of type CallError which would mean it can be used to differentiate between client side and server side errors, right ? @jhendrixMSFT

Reply:
Your assessment is correct.

There's an interesting [blog](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) from Dave Cheney where he asserts "Assert errors for behaviour, not type".  Granted this was written before `errors.As()` and `errors.Is()` were available so perhaps the thinking has somewhat changed, although the advent of those functions don't eliminate the problems he cites.  @element-of-surprise thoughts on this?

Reply:
That is probably my mistake, this looks like an example I gave.  I rarely use errors.Is() or As().  Your assessment seems correct: https://play.golang.org/p/GB8epf1jIov

For our users, they can also simply do it old school:

```go
if ce, ok := err.(CallError); ok { // CallError is always the outer error
  // Do something
}
```

My thoughts tend to go in a similar vein to Dave.  There are two types of errors.
- Ones you can do something about
- Ones you can't

Error values and error types(like CallError) are about errors you can do something about.  fmt.Errorf("") is about ones you can't.  

In my packages I use error values/types when I need to distinguish types that I might be able to do things like retry the call or I need to extract something very specific from the error that I can give the user (I need fields and not just a string).

After that it is strictly about error clarity for the user to manually correct the problem.

Reply:
Thanks for this discussion, updated this part in the latest commit !